### ISSUE TITLE
[Deprecate Dashboard] React error and blank page when trying to disconnect (as non-admin)

### ISSUE THREAD

**Issue by jsnajdr** [Automattic/jetpack#34535]
### Impacted plugin

Jetpack

### Quick summary

Jetpack UI crashes with React error and a blank page when trying to disconnect a non-admin user

### Steps to reproduce

1. Create a self-hosted site, e.g. on Jurassic Ninja, and connect Jetpack
2. Log in as another, non-admin user, and on the `my-jetpack` page connect that user to your WP.com user account
3. Now click "Disconnect" on that page. My expectation was that I'm going to disconnect the account from WP.com, becoming a purely local user again, and maybe have the chance to connect as _another_ user on WP.com.
   <img width="1226" alt="Screenshot 2023-12-08 at 12 53 55" src="https://github.com/Automattic/jetpack/assets/664258/da26c756-cbb3-4d8a-964c-1b927dd0f277">

### A clear and concise description of what you expected to happen.

My local account is disconnected from the WP.com one, and I have an opportunity to connect as another WP.com account.

My intent was _not_ to disconnect the entire site, I'm not an Admin or Owner in the first place. Just the one user account.

### What actually happened

React crashes and I see blank page:

<img width="1081" alt="Screenshot 2023-12-08 at 12 52 52" src="https://github.com/Automattic/jetpack/assets/664258/852c4174-60bc-4b42-8243-e40c08b2a296">

The [React minified error](https://reactjs.org/docs/error-decoder.html/?invariant=31&args[]=%5Bobject%20Error%5D) is:
```
Objects are not valid as a React child (found: [object Error]). If you meant to render a collection of children, use an array instead.
```

### Impact

One

### Available workarounds?

No but the platform is still usable

### Platform (Simple and/or Atomic)

Self-hosted

**Comment by jeherve** [comment]
Thanks for the report! 

This outlines a more general problem with the My Jetpack page for non-admins in general. Should non-admins even have access to that page? They cannot do much there, I don't know if it's really useful to them.

They also will not be able to access any of the product interstitials since the endpoints are limited to admins:
https://github.com/Automattic/jetpack/blob/d4db4c83c59ba0e0895cfff869d682d94591901a/projects/packages/my-jetpack/src/class-rest-products.php#L110-L112

I would be tempted to just limit the pages to admins.

@Automattic/jetpack-agora What's your take on this?

**Comment by jsnajdr** [comment]
I used this screen to connect my local Jetpack site non-admin account to my WP.com one. If I can no longer access it as non-admin, how do I connect then? Does the admin have some workflow to connect me?

**Comment by jeherve** [comment]
You should still have access to the standard Jetpack menu:

<img width="1352" alt="image" src="https://github.com/Automattic/jetpack/assets/426388/97b07976-442c-4a5c-9894-6130484ee93e">

**Comment by jsnajdr** [comment]
Thanks! For a non-admin user, however, the behavior of the Jetpack sidebar menu seems a bit off:
- when I click on the top-level "Jetpack" item, I'm navigated to `?page=my-jetpack` and the "My Jetpack" item is highlighted:

  <img width="175" alt="Screenshot 2023-12-12 at 10 54 31" src="https://github.com/Automattic/jetpack/assets/664258/01fef93a-3051-47bc-8fe1-f351adb56b27">
- when I click on the "Dashboard" item, I'm navigated to `?page=jetpack#/dashboard`. That's the page in your screenshot, where I can connect and disconnect my non-admin account. But the "Dashboard" menu item is not highlighted, only the top-level "Jetpack" is. Although the "Jetpack" menu item link leads elsewhere.
  
  <img width="169" alt="Screenshot 2023-12-12 at 10 58 59" src="https://github.com/Automattic/jetpack/assets/664258/d5445d4d-4dbb-4dd7-934d-e2c1f67fd4ea">

My expectation is that I, as non-admin, don't see "My Jetpack" at all, that "Jetpack" leads to "Dashboard" and that "Dashboard" is highlighted when active.

**Comment by jeherve** [comment]
> My expectation is that I, as non-admin, don't see "My Jetpack" at all, that "Jetpack" leads to "Dashboard" and that "Dashboard" is highlighted when active.

Yes, I think that would be what we should aim for as well.

**Comment by jeherve** [comment]
I took a closer look at this, and was able to reproduce.

The error makes sense at first; the disconnection action makes a call to the `v4/connection` endpoint, which needs the `jetpack_disconnect` permission, set to `manage_options` by default:
https://github.com/Automattic/jetpack/blob/22555bff4bd595dc9821f5217291f571d4bc395c/projects/packages/connection/src/class-manager.php#L1449-L1457

That said, it props a few additional questions:

1. When the user is not an admin, should [the disconnect dialog](https://github.com/Automattic/jetpack/blob/f8d91f73960a86a4ff9c6c2c1795bb9c07f20028/projects/js-packages/connection/components/disconnect-dialog/index.jsx#L14-L20) even try to disconnect the site? Shouldn't it attempt to disconnect just the user making the request, without touching the site connection?
2. If the Disconnect dialog were to behave differently, [the Manage Connection modal](https://github.com/Automattic/jetpack/blob/a1d8fbd4c8a97d1b8181b550aa92e48304467943/projects/js-packages/connection/components/manage-connection-dialog/index.jsx#L93-L98) should probably display different text with different wording when the modal appears for a user connection management request or a site connection management request. We currently only have [one manage connection modal](https://github.com/Automattic/jetpack/blob/de9ee3357b216560dddc101be9d79289422a05f1/projects/packages/my-jetpack/_inc/components/connection-status-card/index.jsx#L146-L157) for both use-cases.


Beyond those questions, **should the My Jetpack page even appear for non-admins?**

- I do not believe it offers any actionable information today, beyond links to "Learn more about" features that do not work.
- Even if those links did work, I'm not sure that information is very valuable to a non-admin who will not be able to activate plugins / features / make decisions about purchases on the site.

Looking at our different plugins today:

1. None of our standalone plugins offer any access to their standalone plugin menus to non-admins, so on those sites only the "My Jetpack" link is displayed.
2. The Jetpack plugin does offer access to its navigation menu because some features can be interesting to non-admins. The Jetpack dashboard offers a way to connect and disconnect your account from WordPress.com though, no need for My Jetpack there.

With that in mind, I would vote for removing My Jetpack entirely for non-admins. It was added in #32420 by @oskosk following a bug described in p1691614027240459-slack-C01U2KGS2PQ. @oskosk Do you think there is still value on showing My Jetpack to non-admins today? Could I try removing it again?

**Comment by oskosk** [comment]
@jeherve My Jetpack was supposed to allow discovery of products and connection problems resoultion. if we restrict to admins , authors are not going to be able to connect from MJ

**Comment by jeherve** [comment]
> if we restrict to admins , authors are not going to be able to connect from MJ

Right, but do they need to be able to connect from My Jetpack? As I mentioned above, what does the user connection give them in the standalone plugins? Jetpack Social seems to be the only standalone plugin where there is value with a user connection, but that's something that can be addressed elsewhere (see #35367).

It has value in the Jetpack plugin, but there we already have an interface (the Jetpack dashboard) where folks can connect (and disconnect) their account.

> discovery of products and connection problems resoultion

These are valuable features, but only to site admins in my opinion, since they're the only ones who can take action and install features / purchase products / solve connection problems.

**Comment by oskosk** [comment]
The Jetpack dashboard is supposed to be replaced by MJ. VideoPress needs a user connection too.

**Comment by jeherve** [comment]
> VideoPress needs a user connection too.

Ah, good point. And although the upsell cards aren't useful to an editor, the connection modal is mandatory so one can actually connect their account (starting from the prompt in the post editor)

<img width="902" alt="Screenshot 2024-02-27 at 18 13 25" src="https://github.com/Automattic/jetpack/assets/426388/6166b708-fad8-40d4-b6e2-db2823246b8f">
<img width="1345" alt="Screenshot 2024-02-27 at 19 05 02" src="https://github.com/Automattic/jetpack/assets/426388/37d78adf-7cf1-4f6d-9092-79552a4ceb7b">

That means we won't be able to just remove the screen for non-admins. We'll need to fix this issue by implementing some of the connection screens suggested in this post:

p1HpG7-hKa-p2

**Comment by CodeyGuyDylan** [comment]
I've opened [a PR that attempted to fix this issue](https://github.com/Automattic/jetpack/pull/40371). It did fix the issue in that the disconnect button now disconnects the user and doesn't throw an error. However now once the user is disconnected, they can no longer reconnect, I keep getting this:

![Image](https://github.com/user-attachments/assets/82c0c01f-c64a-42db-92c7-67f6f8b57085)

It's almost like the disconnection wasn't complete and at least part of the backend thinks the user is still connected. I won't have time to finish this today and I'm off for the next 4 days, so if anybody feels like picking it up where I left off feel free 😄 

Otherwise I will try to get to this early next week

**Comment by CodeyGuyDylan** [comment]
We'll also probably need to update some copy as well as tinker with the connection footer in My Jetpack to hide the `Manage` button if the user is disconnected and they are a non-admin

**Comment by CodeyGuyDylan** [comment]
Tldr: Fixing this issue seems semi-complicated and there are other non-admin related issues present on My Jetpack. Wondering if this should be a project?

@robertsreberski I wonder if this should be upgraded to a project to handle non-admins in My Jetpack. This isn't the only weird thing for non-admins on My Jetpack. For a few examples:

The CTA on the product cards are displayed as clickable links but don't go anywhere:
![Image](https://github.com/user-attachments/assets/24badf33-f38a-4131-acf1-0e5d69e41273)

The recommendations show the regular CTA's when they shouldn't
![Image](https://github.com/user-attachments/assets/210fe46d-8356-4c30-a175-27ca494ef885)

Also, fixing this issue alone would most likely involve updating the connections package to add a new option to only disconnect the user and one to only disconnect the site. Though the problem with that is if the site is disconnected, the user cannot be connected so I guess the options would be "Disconnect User" or "Disconnect Site & User". 

Based on my testing in the PR attached to this issue, there may need to be some backend work done as well as just running the `disconnect_user` function doesn't seem to work because Calypso still thinks the user is logged in when it's been disconnected

I can still move forward and try to solve this issue in isolation if that still sounds like the best option to you

**Comment by robertsreberski** [comment]
Good point @CodeyGuyDylan , thanks for thinking outside of the project 🙏 

We actually have such task planned far in our roadmap: https://github.com/Automattic/jetpack-roadmap/issues/1436

I believe we can add this task there. While it's not urgent, it's quite important - and we will not miss that once the time will come.

In the upcoming months, apart from single maintenance items, let's still focus on boosting quality for admin users, as they are the largest part of our user base. I will prioritize it accordingly, if that sounds good to you.

**Comment by CodeyGuyDylan** [comment]
That sounds good to me 💯

**Comment by jboland88** [comment]
This issue should get resolved by this PR: https://github.com/Automattic/jetpack/pull/41520

**Comment by elliottprogrammer** [comment]
This issue has been fixed in https://github.com/Automattic/jetpack/pull/41520 as part of [[Deprecate Dashboard] My Jetpack for non-admin users](https://github.com/Automattic/jetpack-roadmap/issues/1436#top).

Marking this issue as "closed" (completed) now. ✅
